### ISSUE TITLE
High memory usage and stop working every often

### ISSUE THREAD

**Issue by ThomasdOtreppe** [elastic/elasticsearch#12208]
System info:
- Ubuntu 14.04.2 LTS
- ElasticSearch 1.6.0 from Elastic repository
- 32Gb RAM, 8 CPUs @ 2GHz

I noticed that from time to time, ElasticSearch just stops working and Kibana and Logstash cannot connect anymore to it. It happens every few days (or maybe every day).
When having a closer look at the issue, I noticed that ElasticSearch was still running but stopped listening (no more listening socket) and was using about 1.2Gb of RAM (checked using ps_mem.py). Restarting it fix the issue but it is just a workaround.

Looking at the logs in /var/log/elasticsearch didn't show anything wrong or any warning. There was nothing in system logs either.

Is there a way I can debug that issue? Any flag that I can add to figure out what's going on?

**Comment by caipeichao** [comment]
Maybe you can dump the memory

**Comment by clintongormley** [comment]
Hi @ThomasdOtreppe 

This sounds like a configuration issue.  I suggest you have a read through the setup docs (https://www.elastic.co/guide/en/elasticsearch/reference/current/setup.html) and ask any questions you have in the forum https://discuss.elastic.co/.

If, after that, you think you've found a bug, then feel free to open an issue with the details here.

**Comment by ThomasdOtreppe** [comment]
I already went through the docs. @clintongormley, if I understand correctly, you mean that the listening socket stopping is a configuration option? Weird option.

I'll ask in the forum as you request.

**Comment by clintongormley** [comment]
No, I'm suggesting that you are filling up your heap and probably don't have swap disabled, which is slowing down GCs and making nodes unresponsive.  This is the likeliest condition.

**Comment by ThomasdOtreppe** [comment]
No, the listening socket was gone. netstat -ln didn't show it anymore. It reappeared when I restarted the service.

Have you seen the specs of the system? 32Gb of RAM, 8 CPU @ 2GHz.

**Comment by clintongormley** [comment]
OK I paid more attention to the memory filling up and becoming unresponsive.  I wonder if the socket times out because of slow GCs.  

> Have you seen the specs of the system? 32Gb of RAM, 8 CPU @ 2GHz.

You can fill up 32GB with garbage quite happily, then if you have any heap swapped out, slow GCs will kill responsiveness.  Are you seeing slow GC messages in the logs?  I don't understand why you said "high memory usage" in the title, then say it was only using 1.2 GB?

**Comment by ThomasdOtreppe** [comment]
It wasn't filled. There was maybe a total 2Gb of RAM used according to ps_mem.py. I don't remember seeing any swap used (according to htop) but I could be wrong about that.

A total of 1.7Gb for 2 java processes (Logstash and ElasticSearch). After noticing the issues with ES and killing it, it went down to ~500Mb.

**Comment by clintongormley** [comment]
I wonder if it is related to this futex bug https://github.com/elastic/elasticsearch/issues/11526, https://access.redhat.com/articles/3078

Some other random ideas: 
* what do you set `$ES_HEAP` to?
* If you do `GET /_nodes` do you see `mlockall` set to `true`?
* Are you using the transport client to connect logstash to ES. If so, do you see the same issue if you switch to using the `http` client?
* What do you see in the logs?

**Comment by ThomasdOtreppe** [comment]
I haven't changed the settings in the init.d file which says it defaults to 256Mb, 1Gb max.

I don't see any reference to mlockall when querying /_nodes:
```
{
    "cluster_name": "elasticsearch",
    "nodes": {
        "gmxTUsBZQzCtZU4lpO5x2A": {
            "name": "Doctor Glitternight",
            "transport_address": "local[1]",
            "host": "REDACTED",
            "ip": "PRIVATE_LAN_IP",
            "version": "1.6.0",
            "build": "cdd3ac4",
            "http_address": "inet[/127.0.0.1:9200]",
            "attributes": {
                "local": "true"
            },
            "settings": {
                "node": {
                    "local": "true"
                },
                "config.ignore_system_properties": "true",
                "client": {
                    "type": "node"
                },
                "name": "Doctor Glitternight",
                "pidfile": "/var/run/elasticsearch/elasticsearch.pid",
                "path": {
                    "data": "/var/lib/elasticsearch",
                    "work": "/tmp/elasticsearch",
                    "home": "/usr/share/elasticsearch",
                    "conf": "/etc/elasticsearch",
                    "logs": "/var/log/elasticsearch"
                },
                "config": "/etc/elasticsearch/elasticsearch.yml",
                "cluster": {
                    "name": "elasticsearch"
                },
                "discovery": {
                    "zen": {
                        "ping": {
                            "multicast": {
                                "enabled": "false"
                            }
                        }
                    }
                },
                "network": {
                    "host": "127.0.0.1"
                }
            },
            "os": {
                "refresh_interval_in_millis": 1000,
                "available_processors": 8,
                "cpu": {
                    "vendor": "Intel",
                    "model": "Xeon",
                    "mhz": 2000,
                    "total_cores": 8,
                    "total_sockets": 4,
                    "cores_per_socket": 2,
                    "cache_size_in_bytes": 20480
                },
                "mem": {
                    "total_in_bytes": 33738743808
                },
                "swap": {
                    "total_in_bytes": 34357637120
                }
            },
            "process": {
                "refresh_interval_in_millis": 1000,
                "id": 1912,
                "max_file_descriptors": 65535,
                "mlockall": false
            },
            "jvm": {
                "pid": 1912,
                "version": "1.7.0_79",
                "vm_name": "OpenJDK 64-Bit Server VM",
                "vm_version": "24.79-b02",
                "vm_vendor": "Oracle Corporation",
                "start_time_in_millis": 1436807943854,
                "mem": {
                    "heap_init_in_bytes": 268435456,
                    "heap_max_in_bytes": 1037959168,
                    "non_heap_init_in_bytes": 24313856,
                    "non_heap_max_in_bytes": 224395264,
                    "direct_max_in_bytes": 1037959168
                },
                "gc_collectors": ["ParNew", "ConcurrentMarkSweep"],
                "memory_pools": ["Code Cache", "Par Eden Space", "Par Survivor Space", "CMS Old Gen", "CMS Perm Gen"]
            },
            "thread_pool": {
                "generic": {
                    "type": "cached",
                    "keep_alive": "30s",
                    "queue_size": -1
                },
                "index": {
                    "type": "fixed",
                    "min": 8,
                    "max": 8,
                    "queue_size": "200"
                },
                "fetch_shard_store": {
                    "type": "scaling",
                    "min": 1,
                    "max": 16,
                    "keep_alive": "5m",
                    "queue_size": -1
                },
                "get": {
                    "type": "fixed",
                    "min": 8,
                    "max": 8,
                    "queue_size": "1k"
                },
                "snapshot": {
                    "type": "scaling",
                    "min": 1,
                    "max": 4,
                    "keep_alive": "5m",
                    "queue_size": -1
                },
                "merge": {
                    "type": "scaling",
                    "min": 1,
                    "max": 4,
                    "keep_alive": "5m",
                    "queue_size": -1
                },
                "suggest": {
                    "type": "fixed",
                    "min": 8,
                    "max": 8,
                    "queue_size": "1k"
                },
                "bulk": {
                    "type": "fixed",
                    "min": 8,
                    "max": 8,
                    "queue_size": "50"
                },
                "optimize": {
                    "type": "fixed",
                    "min": 1,
                    "max": 1,
                    "queue_size": -1
                },
                "warmer": {
                    "type": "scaling",
                    "min": 1,
                    "max": 4,
                    "keep_alive": "5m",
                    "queue_size": -1
                },
                "flush": {
                    "type": "scaling",
                    "min": 1,
                    "max": 4,
                    "keep_alive": "5m",
                    "queue_size": -1
                },
                "search": {
                    "type": "fixed",
                    "min": 13,
                    "max": 13,
                    "queue_size": "1k"
                },
                "fetch_shard_started": {
                    "type": "scaling",
                    "min": 1,
                    "max": 16,
                    "keep_alive": "5m",
                    "queue_size": -1
                },
                "listener": {
                    "type": "fixed",
                    "min": 4,
                    "max": 4,
                    "queue_size": -1
                },
                "percolate": {
                    "type": "fixed",
                    "min": 8,
                    "max": 8,
                    "queue_size": "1k"
                },
                "management": {
                    "type": "scaling",
                    "min": 1,
                    "max": 5,
                    "keep_alive": "5m",
                    "queue_size": -1
                },
                "refresh": {
                    "type": "scaling",
                    "min": 1,
                    "max": 4,
                    "keep_alive": "5m",
                    "queue_size": -1
                }
            },
            "network": {
                "refresh_interval_in_millis": 5000,
                "primary_interface": {
                    "address": "PRIVATE_LAN_IP",
                    "name": "eth0",
                    "mac_address": "RE:DA:CT:ED:XX:XX"
                }
            },
            "transport": {
                "bound_address": "local[1]",
                "publish_address": "local[1]",
                "profiles": {}
            },
            "http": {
                "bound_address": "inet[/127.0.0.1:9200]",
                "publish_address": "inet[/127.0.0.1:9200]",
                "max_content_length_in_bytes": 104857600
            },
            "plugins": []
        }
    }
}
```
Here is what I use with logstash:
```
output {
        elasticsearch {
                host => localhost
                protocol => 'http'
        }
}
```

I will try to switch to the http client

As I said, there is nothing unusual in the logs and no warnings or errors. I haven't kept them unfortunately but I'll post them next time it happens.

**Comment by clintongormley** [comment]
It looks like you're already using the http client for logstash.   

You have mlockall turned off (look under the `process` section). Try turning on mlockall (or disabling swap) (see https://www.elastic.co/guide/en/elasticsearch/reference/current/setup-configuration.html#setup-configuration-memory for how) and see if that solves it

Also, are you using a Haswell CPU and can you find out whether your kernel version is susceptible to the futex bug?

**Comment by ThomasdOtreppe** [comment]
How can I find that out? It's a kernel shipped by Ubuntu (3.19.0-22-generic) and even if the base kernel might be susceptible, they might have patched it.

Quick question: is there any good reason using swap is enabled by default? Shouldn't it be disabled by default since it's causing issues?

**Comment by clintongormley** [comment]
> How can I find that out? It's a kernel shipped by Ubuntu (3.19.0-22-generic) and even if the base kernel might be susceptible, they might have patched it.

Then I suggest reading the kernel change logs?  Certainly from my googling attempt, it seems that > 3.18 is OK, but I'm not sure about that.  I'd also check if you have a Haswell processor.

> Quick question: is there any good reason using swap is enabled by default? Shouldn't it be disabled by default since it's causing issues?

Because that is OS level - we can't set that for you.

**Comment by ThomasdOtreppe** [comment]
It's not Haswell according to http://ark.intel.com/products/codename/42174/Haswell#@All : We have Intel(R) Xeon(R) CPU E5-2640 v2 @ 2.00GHz.

Let me rephrase the question. Is there any good reason not to have  'bootstrap.mlockall: true' in the elasticsearch configuration by default since swap causes issues?

**Comment by ThomasdOtreppe** [comment]
Should I increase ES_HEAP_SIZE and if yes, how much?

**Comment by clintongormley** [comment]
> Let me rephrase the question. Is there any good reason not to have 'bootstrap.mlockall: true' in the elasticsearch configuration by default since swap causes issues?

Yes, because we could lock too much memory into RAM, which could bring down your system or starve other processes.  This is not a decision we can take on your behalf.

> Should I increase ES_HEAP_SIZE and if yes, how much?

I refer you to the docs and to the forum again.

**Comment by clintongormley** [comment]
Nothing further - closing
